### PR TITLE
Bump go 1.22.2->1.22.5 and add E2E tests

### PR DESCRIPTION
**Pull request by carreter**
Made some E2E tests that:
- spin up a `kind` cluster
- deploy konnectivity servers and agents
- get metrics to confirm all servers are connected to all agents

Currently do this for two scenarios: 1 agent + 1 server and 3 agents + 3 servers.

Had to run `go mod tidy` and `go mod vendor` along the way to pull in some dependencies, couldn't figure out how to do this without bumping the go patch version. Hope this is ok, I can try and rework it if not!

### CODE REVIEW

**Comment by carreter** [comment]
@avrittrohwer is on vacation this week, is anyone else available to review?

**Comment by carreter** [comment]
Blocks #635 FYI

**Comment by carreter** [comment]
@tallclair for visibility. Working on cleaning up the PR to see if we can avoid upgrading the go version and minimize the vendored dependencies!

**Comment by carreter** [comment]
Seems like `sigs.k8s.io/e2e-fromework v0.4.0` requires go 1.22.3 so it is impossible to downgrade:

```
$ go mod tidy -go 1.22.2
go: sigs.k8s.io/e2e-framework@v0.4.0 requires go@1.22.3, but 1.22.2 is requested
```

Here are the new packages that are getting pulled into `vendor/`:

```
$ git diff master -- vendor/modules.txt | grep -e "+# " | sed -e "s/\+# //" -e "/.*=>.*/d"
github.com/evanphx/json-patch/v5 v5.9.0
github.com/gorilla/websocket v1.5.0
github.com/imdario/mergo v0.3.15
github.com/moby/spdystream v0.2.0
github.com/mxk/go-flowrate v0.0.0-20140419014527-cca7078d478f
github.com/vladimirvivien/gexe v0.2.0
sigs.k8s.io/controller-runtime v0.18.2
sigs.k8s.io/e2e-framework v0.4.0
sigs.k8s.io/yaml v1.4.0
```

@cheftako Where can I check if these are ok to include?

File: go.mod
Comment (carreter) [reply go.mod:4]: `go mod tidy` did this automatically. Reverted.

File: Makefile
Comment (carreter) [reply Makefile:85]: Nope, just didn't know about it. Added the `-race` flag.

File: e2e/templates_test.go
Comment (carreter) [reply e2e/templates_test.go:30]: Sounds like this would be significantly easier to use. Will do!

File: e2e/templates_test.go
Comment (carreter) [reply e2e/templates_test.go:3]: Done!

File: e2e/singleserver_singleagent_test.go
Comment (carreter) [reply e2e/singleserver_singleagent_test.go:25]: Done!

File: e2e/main_test.go
Comment (carreter) [reply e2e/main_test.go:46]: Done!

File: e2e/templates/agent/statefulset.yaml
Comment (carreter) [reply e2e/templates/agent/statefulset.yaml:2]: StatefulSets give the pods a unique identifier we can reach them with. 

```go
metricsFamilies, err := getMetrics(fmt.Sprintf("%v-%v:%v/metrics", agentPod.Name, agentServiceHost, adminPort))
```

File: e2e/main_test.go
Comment (carreter) [reply e2e/main_test.go:114]: Created a `getMetricsGaugeValue()` func

File: e2e/multiserver_multiagent_test.go
Comment (carreter) [reply e2e/multiserver_multiagent_test.go:146]: Done!

**Comment by carreter** [comment]
@avrittrohwer Addressed your comments. Working on including this in the GitHub workflows now.

**Comment by carreter** [comment]
Excluded the e2e tests from the main test runners (fast-test and test) in the Makefile. Hopefully that fixes this.

**Comment by avrittrohwer** [comment]
Hmm, should we expect to see the kind-e2e job in the checks?  I don't see it

**Comment by carreter** [comment]
> Hmm, should we expect to see the kind-e2e job in the checks? I don't see it

I always forget how this works. Might not show up until it gets merged?

File: Makefile
Comment (avrittrohwer) [inline Makefile:70]: do you know if this would exclude any file with e2e in the name?  Could we make this more specific to files within the e2e/ directory

**Comment by avrittrohwer** [comment]
> > Hmm, should we expect to see the kind-e2e job in the checks? I don't see it
> 
> I always forget how this works. Might not show up until it gets merged?

Could you merge this into your fork's main branch then open up a PR again your fork so the new action runs?

File: Makefile
Comment (carreter) [reply Makefile:70]: It would, yes. Updated the pattern to match `/e2e$` and `/e2e/.*` instead.

**Comment by carreter** [comment]
Seems like `kind-e2e` is included as part of the `e2e` workflow but needs maintainer approval before running:

image

There's some minor bugs in the workflow I'm ironing out, will ping when they're sorted!

**Comment by carreter** [comment]
Workflow is working now! I will make it so it runs against the same k8s version matrix as the existing e2e tests.

**Comment by carreter** [comment]
@avrittrohwer Github action for e2e tests is now working against multiple k8s versions. Unfortunately I can't seem to get `make docker-build` to load the image built by the build step, and it rebuilds it for each test run. Any clue how to fix this?

**Comment by avrittrohwer** [comment]
@cheftako @jkh52 could either of you approve the new e2e action check so we can validate this PR?

**Comment by carreter** [comment]
@avrittrohwer FYI, it's working on my fork of the repo: https://github.com/carreter/apiserver-network-proxy/actions/runs/10012999446

**Comment by avrittrohwer** [comment]
> @avrittrohwer Github action for e2e tests is now working against multiple k8s versions. Unfortunately I can't seem to get `make docker-build` to load the image built by the build step, and it rebuilds it for each test run. Any clue how to fix this?

Is this because the test-e2e make target is reliant on docker-build target: `test-e2e: docker-build`?

File: e2e/main_test.go
Comment (ipochi) [inline e2e/main_test.go]: we should reuse kind templates in examples directory. That way we won't have to update templates at two different places. Also it would always mean that if e2e passes.

File: e2e/main_test.go
Comment (ipochi) [inline e2e/main_test.go]: not a blocking change but why not put it together in a config struct ?

File: e2e/singleserver_singleagent_test.go
Comment (ipochi) [inline e2e/singleserver_singleagent_test.go:20]: why change from deployment to statefulset ?

File: e2e/singleserver_singleagent_test.go
Comment (ipochi) [inline e2e/singleserver_singleagent_test.go]: we should also test e2e for http-connect mode.

Perhaps this can be retrieved as environment variable and the e2e github workflow run twice once with grpc and second time with http-connect  ?

similar to how you have done for k8s matrix

File: .github/workflows/e2e.yaml
Comment (ipochi) [inline .github/workflows/e2e.yaml:86]: can you please tell me what this does ? 

Does it not build a new konnectivity server/agent image per PR which is then used by the e2e tests ?

**Review by ipochi — CHANGES_REQUESTED** [review]
Kudos for making the effort to add e2e tests. Left my comments.

I am generally not ok to add new set of templates, I think we should reuse the examples/kind templates the benefit would we two fold.

File: e2e/main_test.go
Comment (carreter) [reply e2e/main_test.go]: How do you mean? I did it this way so it could be passed in easily from the command line (e.g. by the `Makefile`).

File: e2e/singleserver_singleagent_test.go
Comment (carreter) [reply e2e/singleserver_singleagent_test.go:20]: StatefulSets allow for easier addressing of specific pods that belong to a service.

File: e2e/main_test.go
Comment (carreter) [reply e2e/main_test.go]: Good point. In that case, where should the logic for rendering the templates go?

Comment (ipochi) [reply e2e/main_test.go]: I see your point, we add a script in the examples/kind directory that renders  and creates the cluster for folks to try out from the documentation.

but in e2e test, we take the template as is and render and apply in code.

File: e2e/singleserver_singleagent_test.go
Comment (carreter) [reply e2e/singleserver_singleagent_test.go]: Added in the latest commit.

File: e2e/main_test.go
Comment (carreter) [reply e2e/main_test.go]: Exactly. Do you still want me to merge the two of them, or is it ok to keep them separate?

**Comment by carreter** [comment]
Latest commit significantly speeds up the `kind-e2e` runner by not rebuilding the docker images on each iteration of the matrix.

File: e2e/main_test.go
Comment (ipochi) [reply e2e/main_test.go]: we can do that as a separate PR if this is causing blockers. Preference is to use the examples/kind.

Comment (carreter) [reply e2e/main_test.go]: Separate PR would be best as this blocks #635. I'll open an issue.

**Comment by carreter** [comment]
@avrittrohwer @cheftako @ipochi @tallclair how are we feeling about this PR? If there are no major changes to be made, I'll start working on e2e tests for #635. Just want to be sure the e2e framework won't significantly change from under me before I do so!

File: e2e/multiserver_multiagent_test.go
Comment (avrittrohwer) [inline e2e/multiserver_multiagent_test.go:22]: nit: use `replicas` instead of `3` here and below in the agent config

**Review by avrittrohwer — APPROVED** [review]

File: e2e/multiserver_multiagent_test.go
Comment (carreter) [reply e2e/multiserver_multiagent_test.go]: Done!

**Comment by ipochi** [comment]
> @avrittrohwer @cheftako @ipochi @tallclair how are we feeling about this PR? If there are no major changes to be made, I'll start working on e2e tests for #635. Just want to be sure the e2e framework won't significantly change from under me before I do so!

@carreter I'm good.

**Review by ipochi — APPROVED** [review]
Thanks for this effort !!

**Comment by avrittrohwer** [comment]
https://github.com/kubernetes/test-infra/pull/33097 is merged

File: .github/workflows/e2e.yaml
Comment (cheftako) [reply .github/workflows/e2e.yaml:88]: @aojea Could I get you to take a look at this?

Comment (carreter) [reply .github/workflows/e2e.yaml:88]: I was able to fix this by making a new entry in the `Makefile` that doesn't have a dependency on `docker-build`.

File: e2e/singleserver_singleagent_test.go
Comment (cheftako) [reply e2e/singleserver_singleagent_test.go:20]: With Imran here. Agent should not need to maintain state across instances. So from a Kubernetes perspective it should be stateless and hence should run in a deployment not a statefulset. Worried that running it in statefulset in a test could mask a problem.

Comment (carreter) [reply e2e/singleserver_singleagent_test.go:20]: Hmm, do you know of a way to get metrics from a specific pod? That's the reason I was using a StatefulSet.

**Comment by carreter** [comment]
As per @cheftako and @ipochi 's suggestion, I have changed the StatefulSet out for a Deployment and address the pods directly by their IP.

File: e2e/main_test.go
Comment (avrittrohwer) [inline e2e/main_test.go:1]: Could you add a short README.md to the e2e directory that goes over what maint_test.go sets up (RBAC and service) and what each specific test is expected to set up?

Comment (carreter) [reply e2e/main_test.go:1]: On it!

Comment (carreter) [reply e2e/main_test.go:1]: Done, could you take a look?

File: e2e/README.md
Comment (avrittrohwer) [inline e2e/README.md:1]: nit: I don't see 'KinD' capitalization used on the linked docs, keep it 'kind' for consistency

File: e2e/main_test.go
Comment (avrittrohwer) [reply e2e/main_test.go:1]: LGTM!

File: go.mod
Comment (jkh52) [inline go.mod:24]: pleaser restore v0.0.0 to avoid confusion. Note there is a replace directive below.

Comment (carreter) [reply go.mod]: Done!

File: e2e/metrics_assertions_test.go
Comment (cheftako) [inline e2e/metrics_assertions_test.go:24]: Not critical but I think the defer should really be right after the http.Get error check.

File: e2e/metrics_assertions_test.go
Comment (cheftako) [inline e2e/metrics_assertions_test.go:19]: Debuggability is easier if you include the url in with the error.

File: e2e/metrics_assertions_test.go
Comment (cheftako) [inline e2e/metrics_assertions_test.go:69]: Again debuggability goes up if you have a reference to relevant request (server pods) with the error.

File: e2e/metrics_assertions_test.go
Comment (cheftako) [inline e2e/metrics_assertions_test.go:75]: nout found is the obvious error but there could be other reasons we got an error here and we did not check the  error type. I would suggest logging it.

File: e2e/metrics_assertions_test.go
Comment (cheftako) [inline e2e/metrics_assertions_test.go:79]: %v can mask some interesting errors in a case like this. They may be unequal because they are of different types but their toString method generates the same string. If you believe they should both be ints then it is better to use %d.

File: e2e/static_count_test.go
Comment (cheftako) [inline e2e/static_count_test.go:88]: Not wrong but its not clear to me why your type converting what appears to be a constant?

File: e2e/metrics_assertions_test.go
Comment (carreter) [reply e2e/metrics_assertions_test.go:24]: Done.

File: e2e/metrics_assertions_test.go
Comment (carreter) [reply e2e/metrics_assertions_test.go:19]: Done.

File: e2e/metrics_assertions_test.go
Comment (carreter) [reply e2e/metrics_assertions_test.go:69]: Added a reference to the label selector being used to discover the pods.

File: e2e/metrics_assertions_test.go
Comment (carreter) [reply e2e/metrics_assertions_test.go:75]: Done.

File: e2e/metrics_assertions_test.go
Comment (carreter) [reply e2e/metrics_assertions_test.go:79]: Done.

File: e2e/static_count_test.go
Comment (carreter) [reply e2e/static_count_test.go:88]: Compiler complains that adminPort is an int if I don't conver it.

**Pull request merged**
